aiohttp/web_response.py

```diff
@@ -364,7 +364,25 @@ async def _prepare_headers(self) -> None:
         if self._compression:
             await self._start_compression(request)
 
-        if self._chunked:
+        if self.status in (204, 304) or 100 <= self.status < 200:
+            #
+            # Remove transfer-encoding/content-length since there is no
+            # body and this can confuse some clients (e.g. older aiohttp)
+            #
+            # https://datatracker.ietf.org/doc/html/rfc9112#section-6.3
+            #
+            # Any response to a HEAD request and any response with
+            # a 1xx (Informational), 204 (No Content), or 304
+            # (Not Modified) status code is always terminated by
+            # the first empty line after the header fields,
+            # regardless of the header fields present in the message,
+            # and thus cannot contain a message body or trailer section.
+            #
+            if hdrs.TRANSFER_ENCODING in headers:
+                del headers[hdrs.TRANSFER_ENCODING]
+            if hdrs.CONTENT_LENGTH in headers:
+                del headers[hdrs.CONTENT_LENGTH]
```

Comment:
A quick test suggests we don't do the last 2 points currently.

The body is probably a particular issue. If we have a keep-alive connection and we send a body, a conforming client should stop before reading the body, and will then try to interpret the body as a new HTTP message.

Reply:
TODO:

- [ ] At line 407, it correctly removes the Content-Length header if needed already (though could do with checking the full range of 1xx codes).

- [ ] At line 391, I think we just want to make enable_chunking() conditional on the status code. Then pretty much the same as line 407, delete the Transfer-Encoding header for the given status codes (maybe this part can be moved out a level and deduplicated).

- [ ] This is also supposed to be done for a CONNECT request, regardless of status code. Not sure if we have any code for that (or if the method is available at this point in the code).

- [ ] We should probably also make sure that any body provided by the user for these status codes results in an exception, if we don't do this already.

Reply:
>  We should probably also make sure that any body provided by the user for these status codes results in an exception, if we don't do this already.

That seems like it might be better in another PR.

Reply:
> > We should probably also make sure that any body provided by the user for these status codes results in an exception, if we don't do this already.
> 
> That seems like it might be better in another PR.

Up to you, but the title of the PR says that it ensures there is an empty body...

Reply:
It will be a bit more refactoring but I can do it here as well

Reply:
I'm not sure we can do that as these tests fail when I do

```
FAILED tests/test_client_functional.py::test_HTTP_304_WITH_BODY[pyloop] - aiohttp.client_exceptions.ServerDisconnectedError: Server disconnected
FAILED tests/test_client_functional.py::test_HTTP_200_OK_METHOD[pyloop] - aiohttp.client_exceptions.ServerDisconnectedError: Server disconnected
FAILED tests/test_client_functional.py::test_HTTP_200_OK_METHOD_connector[pyloop] - aiohttp.client_exceptions.ServerDisconnectedError: Server disconnected
FAILED tests/test_client_functional.py::test_HTTP_302_REDIRECT_HEAD[pyloop] - aiohttp.client_exceptions.ServerDisconnectedError: Server disconnected
FAILED tests/test_client_functional.py::test_shortcuts[pyloop] - aiohttp.client_exceptions.ServerDisconnectedError: Server disconnected
```

Reply:
Ah, OK, looking at that first test, it does ensure we end up with an empty body, even if a non-empty body is passed in. So, I guess we can leave it like that... Could maybe just tweak one of your tests (if it's not already done) to include a body as input, and assert there is no body as output (to ensure we cover all the status codes).

Reply:
I think its covered by the keep alive tests but I added an explicit one as well

f9e831a0bb1248537852be21b97894e3d21285ce